### PR TITLE
[String] Add support for emoji in AsciiSlugger

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

Usage

```php
use Symfony\Component\String\Slugger\AsciiSlugger;

$slugger = new AsciiSlugger();
$slugger = $slugger->withEmoji();

$slugger->slug('a 😺, 🐈‍⬛, and a 🦁 go to 🏞️... 😍 🎉 💛', '-', 'en');
// a-grinning-cat-black-cat-and-a-lion-go-to-national-park-smiling-face-with-heart-eyes-party-popper-yellow-heart

$slugger->slug('un 😺, 🐈‍⬛, et un 🦁 vont au 🏞️');
// un-chat-qui-sourit-chat-noir-et-un-tete-de-lion-vont-au-parc-national
```

### CODE REVIEW

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php [hunk 1]
```diff
@@ -88,6 +90,14 @@ public function getLocale(): string
         return $this->defaultLocale;
     }
 
+    public function enableEmoji(bool $enableEmoji = true): void
+    {
+        if (!class_exists(EmojiTransliterator::class)) {
+            throw new \LogicException('You cannot use the "Symfony\Component\String\Slugger\AsciiSlugger::enableEmoji()" as the "symfony/int" package is not installed. Try running "composer require symfony/int".');
```
Comment (welcoMattic) [inline src/Symfony/Component/String/Slugger/AsciiSlugger.php]:
```suggestion
            throw new \LogicException('You cannot use the "Symfony\Component\String\Slugger\AsciiSlugger::enableEmoji()" as the "symfony/intl" package is not installed. Try running "composer require symfony/intl".');
```

File: src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php [hunk 2]
```diff
@@ -44,4 +44,27 @@ public function testSlug(string $expected, string $string, string $separator = '
 
         $this->assertSame($expected, (string) $slugger->slug($string, $separator, $locale));
     }
+
+    public function provideSlugEmojiTests(): iterable
+    {
+        yield [
+            'un-chat-qui-sourit-chat-noir-et-a-tete-de-lion-vont-au-parc-national',
```
Comment (welcoMattic) [inline src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php]:
```suggestion
            'un-chat-qui-sourit-chat-noir-et-une-tete-de-lion-vont-au-parc-national',
```

File: src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php [hunk 3]
```diff
@@ -44,4 +44,27 @@ public function testSlug(string $expected, string $string, string $separator = '
...
+            'un 😺, 🐈‍⬛, et a 🦁 vont au 🏞️',
```
Comment (welcoMattic) [inline src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php]:
```suggestion
            'un 😺, 🐈‍⬛, et une 🦁 vont au 🏞️',
```

Comment (lyrixx) [reply src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php]: To, me it's a lion, not a "head of lion". that why I choose to use `un` and not `une`.

Thanks for finding the typo anyway :)

**Comment by lyrixx** [comment]
@welcoMattic Thanks for the review, I addressed your comments

**Review by derrabus — APPROVED** [review]

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php [hunk 4]
```diff
@@ -88,6 +90,14 @@ public function getLocale(): string
         return $this->defaultLocale;
     }
 
+    public function enableEmoji(bool $enableEmoji = true): void
+    {
+        if (!class_exists(EmojiTransliterator::class)) {
```
Comment (stof) [inline src/Symfony/Component/String/Slugger/AsciiSlugger.php]: this should be `if ($enableEmoji && !class_exists(EmojiTransliterator::class))`. There is no requirement if we don't actually enable emoji support.

Comment (94noni) [reply src/Symfony/Component/String/Slugger/AsciiSlugger.php]: Can it be also usefull to add it in « suggest » in the composer.json ?

File: src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php [hunk 5]
```diff
@@ -44,4 +44,27 @@ public function testSlug(string $expected, string $string, string $separator = '
 
         $this->assertSame($expected, (string) $slugger->slug($string, $separator, $locale));
     }
+
+    public function provideSlugEmojiTests(): iterable
+    {
+        yield [
+            'un-chat-qui-sourit-chat-noir-et-un-tete-de-lion-vont-au-parc-national',
+            'un 😺, 🐈‍⬛, et un 🦁 vont au 🏞️',
```
Comment (94noni) [inline src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php:66]: PR desc need to be updated, the « fr » seems wrong

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php (hunk 4 above)
Comment (fabpot) [reply src/Symfony/Component/String/Slugger/AsciiSlugger.php]: No, we’re not doing it anymore.

**Comment by nicolas-grekas** [comment]
I'd prefer going with a wither to keep the service immutable.

File: src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php (hunk 5 above)
Comment (lyrixx) [reply src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php:66]: What does seems wrong ?

Comment (94noni) [reply src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php:66]: et a 🦁 vont au
VS
et un 🦁 vont au

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php (hunk 4 above)
Comment (94noni) [reply src/Symfony/Component/String/Slugger/AsciiSlugger.php]: Oki

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php [hunk 6]
```diff
@@ -88,6 +90,14 @@ public function getLocale(): string
...
+            throw new \LogicException(sprintf('You cannot use the "%s()" as the "symfony/intl" package is not installed. Try running "composer require symfony/intl".', __METHOD__));
```
Comment (gharlan) [inline src/Symfony/Component/String/Slugger/AsciiSlugger.php]:
```suggestion
            throw new \LogicException(sprintf('You cannot use the "%s()" method as the "symfony/intl" package is not installed. Try running "composer require symfony/intl".', __METHOD__));
```

File: src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php (hunk 5 above)
Comment (lyrixx) [reply src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php:66]: Oh, the **description**, got it.

**Comment by lyrixx** [comment]
Hi, I addressed all comments. Thanks for the review

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php [hunk 7]
```diff
@@ -103,6 +117,10 @@ public function slug(string $string, string $separator = '-', string $locale = n
             $transliterator = (array) $this->createTransliterator($locale);
         }
 
+        if ($this->enableEmoji) {
+            $transliterator[] = EmojiTransliterator::create($locale);
```
Comment (lyrixx) [inline src/Symfony/Component/String/Slugger/AsciiSlugger.php]: I have a design question here

On this method (`slug`) the `$locale` could be `null`.
But the locale is mandatory in the `EmojiTransliterator`.

1. Should we check here that the locale is not null?
2. Should we disable the use the transl iterator if `$locale === null`
3. Should we keep the code as it

Anyway, some error could still occurs if someone pass an invalid locale

Comment (ro0NL) [reply src/Symfony/Component/String/Slugger/AsciiSlugger.php]: IMHO 2, as create() doesnt accept null.

Also create() can return null, which AFAIK should be accounted for here.

Comment (lyrixx) [reply src/Symfony/Component/String/Slugger/AsciiSlugger.php]: Done ✅

**Comment by ro0NL** [comment]
Related to #47263, does this mean we need to provision slugger with "en_github/slack" locale? Im not sure it's legit.

**Comment by lyrixx** [comment]
It's not legit IMHO, since the output of the transliterator may contains `:`

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php [hunk 8]
```diff
@@ -88,6 +90,18 @@ public function getLocale(): string
         return $this->defaultLocale;
     }
 
+    public function withEmojiSupport(bool $enableEmoji = true): static
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/Slugger/AsciiSlugger.php]: just `withEmoji(bool $emoji = true)`?
also, what about making it `withEmoji(bool|string $emoji = true)` to allow passing the emoji "locale" when the slugger's locale doesn't exactly match the emoji's? (eg "slack")

File: src/Symfony/Component/String/Slugger/AsciiSlugger.php [hunk 9]
```diff
@@ -103,6 +117,12 @@ public function slug(string $string, string $separator = '-', string $locale = n
             $transliterator = (array) $this->createTransliterator($locale);
         }
 
+        if ($this->enableEmoji && null !== $locale) {
+            if (null !== $tr = EmojiTransliterator::create($locale)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/Slugger/AsciiSlugger.php]: create() will throw when the locale doesn't exist, it won't return null

File: src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php [hunk 10]
```diff
@@ -44,4 +44,32 @@ public function testSlug(string $expected, string $string, string $separator = '
...
+            'en',
+        ];
+        yield [
+            'a-and-a-go-to',
+            'a 😺, 🐈‍⬛, and a 🦁 go to 🏞️... 😍 🎉 💛',
+            null,
+        ];
+    }
+
+    /** @dataProvider provideSlugEmojiTests */
```
Comment (nicolas-grekas) [inline src/Symfony/Component/String/Tests/Slugger/AsciiSluggerTest.php]: CS: should be on many lines + the provider should be after the test method

**Comment by lyrixx** [comment]
@nicolas-grekas I addressed your comments

**Review by nicolas-grekas — APPROVED** [review]

**Comment by ro0NL** [comment]
should we update https://github.com/symfony/symfony/blob/68fe15313443fe1e54459a2060691666fe76830a/src/Symfony/Component/Intl/Transliterator/EmojiTransliterator.php#L52 with soft `@return` to convey it's not a nullable?

**Comment by nicolas-grekas** [comment]
We can remove the `?` actually. Up for a PR @ro0NL?

**Comment by nicolas-grekas** [comment]
(deps=low need a bump for intl in composer.json)

**Comment by lyrixx** [comment]
I fixed requirements for this PR, and also the return type of EmojiTransliterator while I had the code open / cc @ro0NL

**Comment by nicolas-grekas** [comment]
Thank you @lyrixx.

**Pull request merged**